File: leetcode/783.minimum-distance-between-bst-nodes.py
#!/bin/python3
# -*- coding: utf-8 -*-
'''
leetcode #783 minimum-distance-between-bst-nodes 二叉搜索树结点最小距离
https://leetcode-cn.com/problems/minimum-distance-between-bst-nodes/description/
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    MAX_NUM = float('inf')
    l = []
    def addNode(self, node):
        if not node:
            return
        self.l.append(node.val)
        self.addNode(node.left)
        self.addNode(node.right)

    def minDiffInBST(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.l = []
        minDiff = self.MAX_NUM
        self.addNode(root)
        self.l.sort()
        for i in range(len(self.l) - 1):
            minDiff = min(minDiff, self.l[i + 1] - self.l[i])
        return minDiff

File: leetcode/783.minimum-distance-between-bst-nodes.go
package main

import (
	"math"
	"sort"
)

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func minDiffInBST(root *TreeNode) int {
	valSlice := []int{}
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		head := queue[0]
		queue = queue[1:len(queue)]
		valSlice = append(valSlice, head.Val)
		if head.Left != nil {
			queue = append(queue, head.Left)
		}
		if head.Right != nil {
			queue = append(queue, head.Right)
		}
	}
	sort.Ints(valSlice)
	ans := math.MaxInt32
	for i := 1; i < len(valSlice); i++ {
		ans = min(ans, valSlice[i]-valSlice[i-1])
	}
	return ans
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}